kokkos-kernels@4.0.00 +cuda cuda_arch=90 fails build: Kokkos_Cuda_Parallel_Team.hpp(484): error: calling a constexpr __host__ function("operator()") from a __device__ function("exec_team") is not allowed

### Steps to reproduce the issue

`kokkos-kernels@4.0.00 +cuda cuda_arch=90` fails to build using:
* spack@develop (5c6c3b403bc4a8cd08b4e9303748933f3f79f5f7 from `Sun Jun 18 07:17:19 2023 -0400`) - with cherry-pick of https://github.com/spack/spack/pull/36655/commits/5115bcf93023b6dd7e395d51caa5c233b8fba008 to bring in kokkos-kernels update from https://github.com/spack/spack/pull/36655
* GCC 11.1.0
* CUDA 11.8.0
* Ubuntu 20.04, X86_64

Concrete spec: [kokkos-kernels.spec.json.txt](https://github.com/spack/spack/files/11871368/kokkos-kernels.spec.json.txt)

<details><summary>Concretization</summary>
<pre>
 -   kokkos-kernels@4.0.00%gcc@11.1.0~blas~cblas~cublas+cuda~cusparse~ipo~lapack~lapacke~mkl~openmp~serial+shared~superlu~threads build_system=cmake build_type=Release cuda_arch=90 execspace_cuda=auto execspace_openmp=auto execspace_serial=auto execspace_threads=auto generator=make layouts=left memspace_cudaspace=auto memspace_cudauvmspace=auto offsets=int,size_t ordinals=int scalars=double arch=linux-ubuntu20.04-x86_64
[+]      ^cmake@3.26.3%gcc@11.1.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64
[+]          ^ncurses@6.3%gcc@11.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^pkgconf@1.9.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]          ^openssl@1.1.1u%gcc@11.1.0~docs~shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64
[+]              ^ca-certificates-mozilla@2023-01-10%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]              ^perl@5.36.0%gcc@11.1.0+cpanm+open+shared+threads build_system=generic arch=linux-ubuntu20.04-x86_64
[+]                  ^berkeley-db@18.1.40%gcc@11.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64
[+]                  ^bzip2@1.0.8%gcc@11.1.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64
[+]                      ^diffutils@3.9%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                  ^gdbm@1.23%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                      ^readline@8.2%gcc@11.1.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64
[+]              ^zlib@1.2.13%gcc@11.1.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64
[+]      ^cuda@11.8.0%gcc@11.1.0~allow-unsupported-compilers~dev build_system=generic arch=linux-ubuntu20.04-x86_64
[+]          ^libxml2@2.10.3%gcc@11.1.0~python build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^libiconv@1.17%gcc@11.1.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]              ^xz@5.4.1%gcc@11.1.0+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]      ^gmake@4.4.1%gcc@11.1.0~guile build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]      ^kokkos@4.0.00%gcc@11.1.0~aggressive_vectorization~compiler_warnings+cuda~cuda_constexpr~cuda_lambda~cuda_ldg_intrinsic~cuda_relocatable_device_code~cuda_uvm~debug~debug_bounds_check~debug_dualview_modify_check~deprecated_code~examples~hpx~hpx_async_dispatch~hwloc~ipo~memkind~numactl~openmp~openmptarget~pic~rocm+serial+shared~sycl~tests~threads~tuning+wrapper build_system=cmake build_type=Release cuda_arch=90 generator=make intel_gpu_arch=none std=17 arch=linux-ubuntu20.04-x86_64
[+]          ^kokkos-nvcc-wrapper@3.2.00%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
</pre>
</details>

Error:
```
$> spack install -f ./kokkos-kernels.spec.json
...
==> Installing kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x
==> No binary for kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x found: installing from source
==> Fetching https://github.com/kokkos/kokkos-kernels/archive/4.0.00.tar.gz
==> No patches needed for kokkos-kernels
==> kokkos-kernels: Executing phase: 'cmake'
==> kokkos-kernels: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    '/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/gmake-4.4.1-lp2ol6q2quzz7sjtfkrxeczlsvzukkzk/bin/make' '-j16'

21 errors found in build log:
     411    [ 83%] Building CXX object CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTL
            EFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o
     412    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-nvcc-wrapper-3.2.00-ppam7iw5adjksb22f7zn6nmo2ea5xpr6/bin/nvcc_wrapper -DKOKKOS_DEPENDENCE -Dkokkoskernels_EXPORTS -I/tmp/roo
            t/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/common -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7
            ab6hpxwew6vr2x/spack-src/common -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/batched -I/tmp/root/spack-stage/spack-sta
            ge-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256af
            ma/blas -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyj
            nsdbm7ab6hpxwew6vr2x/spack-build-256afma/graph -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph -I/tmp/root/spack-stage/spack-
            stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/co
            mmon/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/src/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-
            256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/src/tpls -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/unit_test -I/tmp/
            root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/dense/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjns
            dbm7ab6hpxwew6vr2x/spack-src/batched/dense/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/dense/unit_test -I/tmp/root
            /spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/sparse/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm
            7ab6hpxwew6vr2x/spack-src/batched/sparse/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/sparse/unit_test -I/tmp/root/
            spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6
            vr2x/spack-src/blas/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernel
            s-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/blas/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/tpls -I
            /tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm
            7ab6hpxwew6vr2x/spack-src/graph/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph/eti -I/tmp/root/spack-stage/spack-stage-
            kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/graph/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-sr
            c/sparse/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-
            256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/sparse/eti -I/tmp/r
            oot/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/tpls -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-4.0.00-6ov
            6e26sbudfqluhfzgfc6omz7abbm4v/include -O3 -DNDEBUG -std=c++17 -fPIC -arch=sm_90 -MD -MT CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_symbolic/Sparse_p
            ar_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o -MF CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_i
            lut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o.d -o CMakeFiles/kokkoskernels.dir/sparse/eti/generated_spe
            cializations_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o -c /tmp/root/spack-stage/spack-stage
            -kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/sparse/eti/generated_specializations_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_O
            FFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp
     413    [ 83%] Building CXX object CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTL
            EFT_EXECSPACE_SERIAL_MEMSPACE_HOSTSPACE.cpp.o
     414    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-nvcc-wrapper-3.2.00-ppam7iw5adjksb22f7zn6nmo2ea5xpr6/bin/nvcc_wrapper -DKOKKOS_DEPENDENCE -Dkokkoskernels_EXPORTS -I/tmp/roo
            t/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/common -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7
            ab6hpxwew6vr2x/spack-src/common -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/batched -I/tmp/root/spack-stage/spack-sta
            ge-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256af
            ma/blas -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyj
            nsdbm7ab6hpxwew6vr2x/spack-build-256afma/graph -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph -I/tmp/root/spack-stage/spack-
            stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/co
            mmon/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/src/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-
            256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/src/tpls -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/unit_test -I/tmp/
            root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/dense/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjns
            dbm7ab6hpxwew6vr2x/spack-src/batched/dense/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/dense/unit_test -I/tmp/root
            /spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/sparse/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm
            7ab6hpxwew6vr2x/spack-src/batched/sparse/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/sparse/unit_test -I/tmp/root/
            spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6
            vr2x/spack-src/blas/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernel
            s-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/blas/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/tpls -I
            /tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm
            7ab6hpxwew6vr2x/spack-src/graph/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph/eti -I/tmp/root/spack-stage/spack-stage-
            kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/graph/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-sr
            c/sparse/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-
            256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/sparse/eti -I/tmp/r
            oot/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/tpls -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-4.0.00-6ov
            6e26sbudfqluhfzgfc6omz7abbm4v/include -O3 -DNDEBUG -std=c++17 -fPIC -arch=sm_90 -MD -MT CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_symbolic/Sparse_p
            ar_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_SERIAL_MEMSPACE_HOSTSPACE.cpp.o -MF CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par
            _ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_SERIAL_MEMSPACE_HOSTSPACE.cpp.o.d -o CMakeFiles/kokkoskernels.dir/sparse/eti/generated
            _specializations_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_SERIAL_MEMSPACE_HOSTSPACE.cpp.o -c /tmp/root/spack-stage/spack
            -stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/sparse/eti/generated_specializations_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL
            _INT_OFFSET_SIZE_T_LAYOUTLEFT_EXECSPACE_SERIAL_MEMSPACE_HOSTSPACE.cpp
     415    [ 84%] Building CXX object CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_numeric/Sparse_par_ilut_numeric_eti_DOUBLE_ORDINAL_INT_OFFSET_INT_LAYOUTLEFT_E
            XECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o
     416    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-nvcc-wrapper-3.2.00-ppam7iw5adjksb22f7zn6nmo2ea5xpr6/bin/nvcc_wrapper -DKOKKOS_DEPENDENCE -Dkokkoskernels_EXPORTS -I/tmp/roo
            t/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/common -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7
            ab6hpxwew6vr2x/spack-src/common -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/batched -I/tmp/root/spack-stage/spack-sta
            ge-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256af
            ma/blas -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyj
            nsdbm7ab6hpxwew6vr2x/spack-build-256afma/graph -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph -I/tmp/root/spack-stage/spack-
            stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/co
            mmon/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/src/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-
            256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/src/tpls -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/common/unit_test -I/tmp/
            root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/dense/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjns
            dbm7ab6hpxwew6vr2x/spack-src/batched/dense/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/dense/unit_test -I/tmp/root
            /spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/sparse/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm
            7ab6hpxwew6vr2x/spack-src/batched/sparse/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/batched/sparse/unit_test -I/tmp/root/
            spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6
            vr2x/spack-src/blas/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernel
            s-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/blas/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/blas/tpls -I
            /tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm
            7ab6hpxwew6vr2x/spack-src/graph/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/graph/eti -I/tmp/root/spack-stage/spack-stage-
            kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/graph/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-sr
            c/sparse/src -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/impl -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-
            256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/eti -I/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/sparse/eti -I/tmp/r
            oot/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-src/sparse/tpls -isystem /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-4.0.00-6ov
            6e26sbudfqluhfzgfc6omz7abbm4v/include -O3 -DNDEBUG -std=c++17 -fPIC -arch=sm_90 -MD -MT CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_numeric/Sparse_pa
            r_ilut_numeric_eti_DOUBLE_ORDINAL_INT_OFFSET_INT_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o -MF CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_n
            umeric/Sparse_par_ilut_numeric_eti_DOUBLE_ORDINAL_INT_OFFSET_INT_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o.d -o CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializatio
            ns_cpp/par_ilut_numeric/Sparse_par_ilut_numeric_eti_DOUBLE_ORDINAL_INT_OFFSET_INT_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o -c /tmp/root/spack-stage/spack-stage-kokkos-kernels
            -4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/sparse/eti/generated_specializations_cpp/par_ilut_numeric/Sparse_par_ilut_numeric_eti_DOUBLE_ORDINAL_INT_OFFSET_INT_LAYOUTL
            EFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp
  >> 417    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-4.0.00-6ov6e26sbudfqluhfzgfc6omz7abbm4v/include/Cuda/Kokkos_Cuda_Parallel_Team.hpp(484): error: calling a constexpr __host__
             function("operator()") from a __device__ function("exec_team") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
     418              detected during:
     419                instantiation of "std::enable_if_t<std::is_void<TagType>::value, void> Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::exec_team<Ta
            gType>(const Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::Member &) const [with FunctorType=lambda [](const member_type &)->void, Properties
            =<Kokkos::CudaSpace::execution_space>, TagType=Kokkos::Impl::WorkTagTrait::base_traits::work_tag]"
     420    (511): here
     421                instantiation of "void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::operator()() const [with FunctorType=lambda [](const member_
            type &)->void, Properties=<Kokkos::CudaSpace::execution_space>]"
     422    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/kokkos-4.0.00-6ov6e26sbudfqluhfzgfc6omz7abbm4v/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp(89): here
     423                instantiation of "void Kokkos::Impl::cuda_parallel_launch_local_memory(DriverType) [with DriverType=Kokkos::Impl::ParallelFor<lambda [](const member_type &)->void, Kokkos::
            TeamPolicy<Kokkos::CudaSpace::execution_space>, Kokkos::Cuda>]"
...
     459    2 errors detected in the compilation of "/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma/sparse/eti/generated_specializatio
            ns_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_ORDINAL_INT_OFFSET_INT_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp".
  >> 460    make[2]: *** [CMakeFiles/kokkoskernels.dir/build.make:2095: CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_symbolic/Sparse_par_ilut_symbolic_eti_DOUBLE_
            ORDINAL_INT_OFFSET_INT_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o] Error 1
...
  >> 785    make[2]: *** [CMakeFiles/kokkoskernels.dir/build.make:2151: CMakeFiles/kokkoskernels.dir/sparse/eti/generated_specializations_cpp/par_ilut_numeric/Sparse_par_ilut_numeric_eti_DOUBLE_OR
            DINAL_INT_OFFSET_INT_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDASPACE.cpp.o] Error 1
     786    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma'
  >> 787    make[1]: *** [CMakeFiles/Makefile2:113: CMakeFiles/kokkoskernels.dir/all] Error 2
     788    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-kokkos-kernels-4.0.00-256afmajipyjnsdbm7ab6hpxwew6vr2x/spack-build-256afma'
  >> 789    make: *** [Makefile:139: all] Error 2
```

### Error message

<details><summary>Error message</summary>

<pre>
...
</pre></details>


### Information on your system

* **Spack:** 0.21.0.dev0 (10fd13d9ba7d7ec301e62b95f0eca862aafadc91)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/11871371/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/11871372/spack-build-out.txt)

@lucbv @brian-kelley @srajama1 @wspear @nchaimov

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/kokkos-kernels/package.py
+++ b/var/spack/repos/builtin/packages/kokkos-kernels/package.py
@@ -78,6 +78,9 @@ class KokkosKernels(CMakePackage, CudaPackage):
         variant(eti, default=deflt, description=descr)
         depends_on("kokkos+%s" % backend_required, when="+%s" % eti)
 
+    # kokkos-kernels requires KOKKOS_LAMBDA to be available since 4.0.00
+    depends_on("kokkos+cuda_lambda", when="@4.0.00:+cuda")
+
     numeric_etis = {
         "ordinals": (
             "int",

